i18n utility script (#5851)
newrelic/docs-website: scripts/i18n_utility/README.md

# I18n Utility Script

> 

Reviewer: I think we could use an intro paragraph here that provides some larger context. It could cover:

1. How the translation files are created and map to the English files/directory structure
2. What happens when there's a mismatch between the translated directories and the English site content
3. What "I18n" means (I assume this is a funny shorthand for "internationalization" which, ick :D )
4. I'm sure there's something else I'm not considering

> This is a CLI utility for:

Reviewer: I'd change "CLI" to "command-line" here

I'd also change "for" to "to" so that you can so you can use the infinitive (for example, to: identify, to: delete, to: identify, etc)

* identifying orphaned translation related files
* deleting orphaned files
* identifying rename changes to make for translation related files
* making rename changes for translation related files

> Orphaned files refer to files that exist underneath `i18n/content/docs` that do not have an english equivalent and are therefore unused and safe to delete.

Reviewer: ```suggestion
Orphaned files are files in `i18n/content/docs` that don't have an English equivalent. They're unused and safe to delete.
```

> For the rename changes, this refers to when you move an english document and don't move an existing translated version of that doc. So if we have a doc `hello/world.txt` and we rename it to `hello/goodbye.txt` we also want to rename the translated versions of that document (if they exist). The last two commands will print out subsequent rename changes that need to be made (based on what has been renamed already in the english content directory) and make those changes for you.

Reviewer: I was briefly baffled by this paragraph. I'm sure we can clarify and simplify this explanation. Also, why are we using .txt for examples when we work entirely in .mdx files :)

"When you rename a doc's filename, you need to make sure the translated file is renamed to match.

/docs/content/../hello-world.mdx   --> /docs/content/../goodbye-world.mdx  

/docs/content/../i18n/../hello-world.mdx --> /docs/content/../i18n/../goodbye-world.mdx 
"

Or whatever the paths are

Author: just to nit pick, this change isn't limited to mdx files, but everything else i agree with 🙈 

any file that lives under the i18n content directory is a candidate. .txt is probably an inferior example compared to .mdx, but a fair example would be images.

> ## Important Notes

Reviewer: I'd change this header to "Troubleshooting"

There are a couple scenarios that are easy enough to find yourself in where you may not see output that you expect to, or you think the script isn't running correctly.

> ### Scenario: Branch is not up to date

Reviewer: I'd change this to describe the the situation "You see a rename change you didn't make"

It may be the case that you see a (rename) change listed that you did not make. The script runs against `origin/develop` and creates a diff between the two branches. This can include changes that do not exist locally if the branch is behind develop. The more behind a branch is, the higher the chance of running into this is.

The fix: keep working branch up to date. Merge or rebase with develop frequently.

> ### Scenario: No renames detected

Reviewer: "Your rename isn't detected by git"

> If files are manually being moved around, there is a chance that git does not detect that as a rename operation. Instead, git may initially see this as having created & deleted two files. If the script is run in this instance, renamed files will not be detected.

Reviewer: Personally, it'd be great to have a better understanding of why this happens (changes not being registered by git).

Author: if the explanation below is good to you, i can add that to the readme.

> The fix: As much as possible, when renaming files/folders, the `git mv` command should be used. This explicitly tells git the operation is a rename, and the script should work as expected.

Reviewer: Should we really be using `git mv` when renaming files/folders?

Author: it depends. if you dont use git mv, git will naturally see changes as add & delete, until you stage them. at that point, git may see them as rename changes. if you make changes & move a file, git may not be able to detect that you renamed a file.

when you use `git mv` you explicitly tell git that you renamed a file, and the change is staged.
